[WIP] [Type Promotion] Add floating type promotion support for (some of the) Unary Floating UFuncs

krshrimali opened the pull request:
This PR proposes to add type promotion (Int and Bool Type to Default Floating Type) to the Unary Ops which are appropriate for this kind of type promotion (called `Unary Floating UFuncs` here).

**Some context:** (Also see #28703 issue) NumPy follows type promotion for it's Unary Ufuncs as mentioned below:

**Type - 1:** int8 - float16, int16 - float32, int32 - float64, int64 - float64, bool - float16
**Type - 2:** (int8, int16, int32, int64, bool) - float64
**Type - 3:** bool - int8
**Type - 4:** bool - float16

- Ops with no upcasting: abs, real, imag, bitwise_not, clip, neg
- Ops with changes (Type 1): ceil, expm1, floor, log, log10, log1p, log2, sin, sinh, sqrt, trunc, atan, cos, tan
- Ops with changes (Type 2): angle
- Ops with changes (Type 3): conjugate, reciprocal, square
- Ops with changes (Type 4): round

After offline and online discussions with @nairbv, @mruberry, @mcarilli, it was decided that we don't want to promote all the ops listed above (see why: https://github.com/pytorch/pytorch/pull/33322#discussion_r384006479). JAX's NumPy rules for type promotion of Unary UFuncs were also explored, and after consensus, we decided to promote **Int & Bool Tensors to Default Floating Type Tensors** for the following ops:

```
sin, cos, tan, asin, acos, atan, sinh, cosh, tanh, ceil, floor, exp, expm1, log, log10, log1p, log2, sqrt, rsqrt, trunc, erf, erfc, erfinv, lgamma, digamma, sigmoid
```

_Note: This PR does not enable floating type promotion for in-place code paths for obvious reasons._

**Comprehensive list of changes proposed in this PR:**

1. Add floating type promotion (see list of floating unary ufuncs and rules above) to appropriate Unary Universal Functions.
2. Add a test class for Unary Floating UFuncs in `test_torch.py` file. This generates tests for all the unary floating ufuncs. The tests also cover complex types and float to complex type promotion for the appropriate ops.
3. Add tests for Unary Floating UFuncs in `test_pytorch_onnx_onnxruntime.py` file. 

cc: @mruberry, @nairbv, @mcarilli

krshrimali committed:
Allows int to float conversions for Unary Ops

krshrimali committed:
Typo solved

krshrimali committed:
Pass tensor instead of type to the function, some typos resolved

krshrimali commented:
Note: I am working on splitting this for ops which need this promotion and which don't (looks like not _all_ the Unary Ops should have this feature). Adding tests for the modifications, is also in the TODO list.

nairbv commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -49,8 +79,17 @@ static inline Tensor& unary_op_impl_out(Tensor& result, const Tensor& self, Stub
-  Tensor result = at::empty({0}, self.options());
-  return out_impl(result, self);
+  // This enables int-to-float implicit dtype conversions
+  ScalarType promoted_dtype = promoteIntToFloats(self);
```
we'll need a way to differentiate which ops shouldn't promote int->float, e.g. `torch.abs()`

krshrimali replied:
From the checks I did on the numpy ops, there are 4 types of conversions on the Unary Ops:

 1. Type - 1: int8 - float16, int16 - float32, int32 - float64, int64 - float64, bool - float16
 2. Type - 2: (int8, int16, int32, int64, bool) - float64
 3. Type - 3: bool - int8
 4. Type - 4: bool - float16

Also, some ops are not defined in NumPy (like `sigmoid`, `polygamma`, `mvlgamma`). What do you suggest? Should we follow the type promotion similar to NumPy's strategy? Which, in case, will lead us to have conditions for the ops to go through a specific type (1/2/3/4) of dtype promotion.

In case you are interested to know which ops use what type of promotion:

 * Ops with no upcasting: `abs`, `real`, `imag`, `bitwise_not`, `clip`, `neg`
 * Ops with changes (Type 1): `ceil`, `expm1`, `floor`, `log`, `log10`, `log1p`, `log2`, `sin`, `sinh`, `sqrt`, `trunc`, `atan`, `cos`, `tan`
 * Ops with changes (Type 2): `angle`
 * Ops with changes (Type 3): `conjugate`, `reciprocal`, `square`
 * Ops with changes (Type 4): `round`

nairbv replied:
@gchanan do you think we'd be willing to deviate from NumPy on some of these UnaryOp promotions?

0. (not promoting) these seem reasonable.
1. Seems reasonable.
2. This case for one operator (angle) is expecting a complex argument. The angle of any real argument is 0. These ops currently handle (most) integral types without promoting, and I'm not convinced we should change that behavior as part of this ticket.
3. `bool -> int8` for conjugate/reciprocal/square seem a bit weird to me. E.g. I'd expect `t.square()` to be the same as `t * t`. It's also possible we'd want to skip boolean tensors for these operators, since ops like bool_tensor.reciprocal() would crash if they contain any False values.
4. `round` This inconsistent behavior seems odd since none of the other integral types promote for that op. I'd be tempted to deviate from np on this one, or just not implement for bool.

krshrimali committed:
Allowing implicit dtype promotion matching NumPy's strategies

krshrimali commented:
The latest commit is an example code for allowing dtype promotion matching to NumPy on sample functions. Here are the results on sample ops:

```
>>> x = torch.tensor(2, device="cuda")
>>> y = torch.tensor(False, device="cuda") 

>>> torch.abs(x).dtype # No change
torch.int64

>>> torch.ceil(x).dtype # Type 1 (int64 to float64)
torch.float64
>>> torch.angle(x).dtype # Type 2 (int64 to float64)
torch.float64
>>> torch.conj(y).dtype # Type 3 (bool to int8)
torch.int8
>>> torch.round(y).dtype # Type 4 (bool to fp16 if cuda, else fp32)
torch.float16
```

I welcome inputs and suggestions on this prototype. @nairbv @mcarilli

mcarilli commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -31,6 +31,98 @@
 namespace at {
 namespace native {
 
+enum class TypePromotionStrategy {
```
In general, functions that are only used statically (within one file) should be declared + defined within an anonymous namespace, so they don't pollute the global at::native namespace.
```
namespace {
  promoteType1() {
  }
  promoteType2() {
  }
  ....
} // end anonymous namespace

unary_op_impl() {
  promoteType1(...)
}
```

krshrimali replied:
Thanks, @mcarilli. I've now committed changes using anonymous namespace!

krshrimali committed:
Using anonymous namespace, to not pollute at::native global namespace

krshrimali committed:
Add tests for an op of each type

krshrimali commented:
I've added sample test functions for an op of each type/category, in `test_torch.py`. Adding for each op will have a lot of code repetition, and to avoid that, we can have a test function per each category run for all the respective ops. I'm figuring out the best way to do that, will appreciate any inputs.

mcarilli commented on test/test_torch.py:
```diff
@@ -11882,6 +11882,50 @@ def test_ceil_out_mismatch(self, device):
+                self.assertIs(out.dtype, torch.float16)
+
+    @dtypes(torch.int8, torch.int16, torch.int32, torch.int64, torch.bool)
+    def test_angle_type_upcasting(self, device, dtype):
```
To make this generic, have a list of op name strings, and do something like
```def test_type_1(self):
        type_1_list = ["angle", ...]
        for opname in type_1_list:
            getattr(torch, opname)(arg)
```

krshrimali replied:
Thanks, @mcarilli. I've added the generic test functions in the recent commits.

mcarilli commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -31,6 +31,101 @@
 namespace at {
 namespace native {
 
+// Using anonymous namespace as these functions' scope is limited to this file and we do not want to pollute at::native namespace
```
don't need a comment, it's a fairly common pattern across all of pytorch

krshrimali replied:
Okay! Removed the comment :)

krshrimali committed:
code cleaning, generic test functions

krshrimali committed:
Long comments not favored, Lint tests should pass now

krshrimali committed:
keep unary_op_impl short, move dtype calculation to a helper function instead

krshrimali committed:
keeping unary_op_impl short, move dtype calculation to helper function

krshrimali committed:
remove anonymous namespace comment

krshrimali committed:
remove empty lines (nit) from the helper functions

nairbv commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -31,6 +31,122 @@
 namespace native {
 
+namespace {
+  enum class TypePromotionStrategy {
```
I'm hoping we can simplify this, but if we need the whole Enum we probably need more descriptive names vs `TypeN`

krshrimali replied:
Thanks for your inputs @nairbv. We went with the `Enum` approach, as the unary ops have different dtype promotion policies, and we needed an easier way for each op to communicate with the dtype promotion policy. We also wanted to not pollute the helper function `unary_op_impl` with any type promotion strategies, and to keep it short, we use helper functions defined in the anonymous namespace.

I agree, that the names need to be more descriptive. The next commit will resolve that.

krshrimali replied:
I've changed the names to more descriptive names in the Enum. (https://github.com/pytorch/pytorch/pull/33322/commits/2203eb9fae716bad2b5907dfc4518323ca4ee5c5)

nairbv commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -31,6 +31,122 @@
+        dtype = (self.device().type() == DeviceType::CPU) ? kFloat : kHalf;
+        break;
+      case kShort:
+        dtype = kFloat;
```
might be simpler to just put the return statements in each case here instead of saving and breaking to return later.

krshrimali replied:
Agreed. This should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/2203eb9fae716bad2b5907dfc4518323ca4ee5c5

krshrimali committed:
Testing with minor changes...

krshrimali committed:
More descriptive names (vs TypeN) in the enum, minor edits

1. enum TypePromotionStrategy had TypeN format. This has changed to more descriptive names (IntBoolToFloats, IntBoolToFloat64, BoolToInt8, BoolToFloat16) as suggested by @nairbv
2. Functions do not need to store dtypes and then break in switch-case statements, they can directly return dtypes as suggested by @nairbv
3. Extra lines removed, they just take more space. As suggested by @mcarilli

krshrimali committed:
Descriptive names in enum for TypePromotionStrategy, minor edits

This commit:

1. Adds more descriptive names in the Enum for TypePromotionStrategy (earlier TypeN format used), as suggested by @nairbv
2. Returns dtypes directly from switch case (earlier: store, break and return strategy used), as suggested by @nairbv
3. Removes extra lines, as suggested by @mcarilli

Merge branch 'temp-unary-ops' into int-to-float-unary-ops

krshrimali committed:
fix build errors

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

mruberry commented:
Recapping our offline conversation: let's refocus this PR to implement unary floating ufunc type promotion where all integral or bool inputs are promoted to the default floating type. For example, sin(IntTensor)->DefaultFloatTensor.

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
Modify approach to only Int&Bool to (default) float type, added tests

Based on the offline conversation, this PR has shifted it's focus to implement unary floating ufunc type promotion. The approach is to promote integral or boolean inputs to default floating type. Unary Floating Ops (to be added): atan, cos, cosh, tanh, tanh (implemented using different macros)

krshrimali commented on test/test_torch.py:
```diff
@@ -6066,13 +6066,6 @@ def test_neg(self, device):
-            r"Negation, the `\-` operator, on a bool tensor is not supported. "
-            r"If you are trying to invert a mask, use the `\~` or `logical_not\(\)` operator instead.",
-            lambda: - torch.tensor([False, True], device=device))
-
```
This assert was removed since `torch.neg` now supports Boolean tensors (implicitly promoted to default floating type) as inputs. This means: `torch.neg(torch.tensor(False))` now returns `-0.` .

krshrimali committed:
(Corrected list of floating unary ops) - added tests and helper function for type promotion

krshrimali commented on test/test_torch.py:
```diff
@@ -12151,6 +12151,17 @@ def test_ceil_out_mismatch(self, device):
+            x = torch.tensor([2], dtype=dtype, device=device)
+            out = getattr(torch, op_name)(x)
+            # Assert all int and bool dtypes are promoted to default floating dtype
+            self.assertIs(out.dtype, torch.get_default_dtype())
```
This adds generic test for the floating unary ops. There are a few ops left to integrate with this type promotion PR, and the list in this test code will also be updated in the next commit.

gchanan replied:
seems good to check that the default dtype is actually respected, i.e. test both float and double default dtypes.

krshrimali replied:
Thank you for the suggestion @gchanan , I'll add that in the next commit(s).

krshrimali replied:
This should be solved here (https://github.com/pytorch/pytorch/pull/33322/commits/df7f5c60b3d9056c9d35c9e50fa5f56fc011b00d). Please let me know if it doesn't, or if you have any more suggestions. Thanks! :)

krshrimali replied:
This has been made simpler by using `@float_double_default_dtype`.

krshrimali committed:
Lint annotation failure solved

krshrimali committed:
Add macros (IMPLEMENT_UNARY_FLOATING_OP_*) for the remaining ops

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,17 +53,26 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
+// See PR #33322 for more discussion on this
+template <typename OutImpl>
+static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl) {
+  Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : \
```
The linter will likely complain about line length. You can break this line at the "?".

krshrimali replied:
I agree, I should have done this at the first place. Thanks!

mcarilli replied:
semi-related tip:  I know you like vim.  `/ $` is helpful to highlight/catch trailing whitespace.

krshrimali replied:
Thanks @mcarilli. :+1:

krshrimali replied:
Removed white-space and added break line! (https://github.com/pytorch/pytorch/pull/33322/commits/c816f39f45369536c2ffb329756ae36405b21d64)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,17 +53,26 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
+static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl) {
+  Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : \
+                  at::empty({0}, self.options());
+  return out_impl(result, self.to(result.scalar_type()));
```
Is this cast of self necessary? What happens without it?

krshrimali replied:
I think it's necessary: `torch.expm1(torch.tensor(2))` throws `RuntimeError: expected dtype Float but got dtype Long` error.

krshrimali replied:
It's probably because of this check here: https://github.com/pytorch/pytorch/blame/master/aten/src/ATen/native/TensorIterator.cpp#L143

mcarilli replied:
I think it will create a TensorIterator::unary_op with mismatched types, and because TensorIterator::unary_op doesn't yet support internal promotion, that will error.  That's part of what we have to think about for the next stage (promotion based on an explicit dtype argument).

mcarilli replied:
lol, i was writing my comment on a stale version of the webpage...

krshrimali replied:
This should be resolved here. https://github.com/pytorch/pytorch/pull/33322/commits/c2ff01ba86bc13ce422c8fb49126aabf547b2b29

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,17 +53,26 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
   return out_impl(result, self);
 }
 
+// Implicitly promote all int and bool dtypes to default floating dtype
```
Let's put enough into this comment that we don't need to reference the PR and people know when to use this function instead of unary_op_impl. What are the key points?

1. This ensures the output type of the function is a floating point value.
2. It's appropriate for functions like sin() which (mathematically speaking) produce non-integral values.
3. Implementers should continue to use the other unary functions, like unary_op_impl_, since they explicitly specify the output type. 

Does that sound reasonable? Also, the promotion is not implicit (it's very explicit).

krshrimali replied:
Okay, sure! I'll add a detailed comment containing the above key points instead of linking to the PR.

krshrimali replied:
Added the comments! :) (https://github.com/pytorch/pytorch/pull/33322/commits/c816f39f45369536c2ffb329756ae36405b21d64)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -310,6 +319,16 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
     return result;                                                     \
   }
 
+// IMPLEMENT_UNARY_FLOATING_OP_* macros are used for ops which support 
```
Per the above, point to the previous comment, not the PR.

krshrimali replied:
Done! (https://github.com/pytorch/pytorch/pull/33322/commits/c816f39f45369536c2ffb329756ae36405b21d64)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -310,6 +319,16 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
+// Check PR #33322 for more details on this
+#define IMPLEMENT_UNARY_FLOATING_OP_CORE(op)                           \
+  Tensor op(const Tensor& self) {                                      \
+    Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : at::empty({0}, self.options()); \
```
Line too long.

krshrimali replied:
The recent commit (https://github.com/pytorch/pytorch/pull/33322/commits/c816f39f45369536c2ffb329756ae36405b21d64) should solve this! :)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -310,6 +319,16 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
+#define IMPLEMENT_UNARY_FLOATING_OP_CORE(op)                           \
+  Tensor op(const Tensor& self) {                                      \
+    Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : at::empty({0}, self.options()); \
+    at::op##_out(result, self.to(result.scalar_type()));               \
```
Per above, is this cast needed?

krshrimali replied:
This should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/c2ff01ba86bc13ce422c8fb49126aabf547b2b29.

krshrimali committed:
Break line, remove whitespaces, add detailed comments

krshrimali committed:
removed sign from the list, it doesn't need upcasting

krshrimali committed:
Test for both float and double default dtypes, refine ops list & add comment

krshrimali committed:
remove backslashes - for lint testing

mcarilli commented on test/test_torch.py:
```diff
@@ -12151,6 +12151,25 @@ def test_ceil_out_mismatch(self, device):
+                   "tanh"]
+        # Make sure that the ops promote to floating default types
+        # Test both float32 and double (float64) dtypes
+        saved_dtype = torch.get_default_dtype()
```
It might be worth defining a context manager to save and restore the dtype, so the dtype will be restored even if something in the loop below throws an error.

mcarilli replied:
Resolved by https://github.com/pytorch/pytorch/pull/33322/#discussion_r388504876

nairbv commented on test/test_torch.py:
```diff
@@ -12151,6 +12151,25 @@ def test_ceil_out_mismatch(self, device):
 
+    # Test for unary ops appropriate for floating-type output from integral and boolean inputs
+    @dtypes(torch.int8, torch.int16, torch.int32, torch.int64, torch.bool)
+    def test_intbool_to_float_upcasting(self, device, dtype):
```
should this go in test_type_promotion.py?

krshrimali replied:
Sure! I'll move it to test_type_promotion.py

krshrimali replied:
There is a plan to create a Test Class for all the Unary Ufuncs, as per discussion with @mruberry. This test was although supposed to go in `test_type_promotion.py` as correctly suggested by @nairbv, but since we are generalizing it for all the unary ufuncs, so we'll keep it in `test_torch.py`.

nairbv commented on test/test_torch.py:
```diff
@@ -12151,6 +12151,25 @@ def test_ceil_out_mismatch(self, device):
+                out = getattr(torch, op_name)(x)
+                # Assert all int and bool dtypes are promoted to default floating dtype
+                self.assertIs(out.dtype, torch.get_default_dtype())
+        torch.set_default_dtype(saved_dtype)
```
FYI there's a decorator `@float_double_default_dtype` which will run the test twice with each default dtype set. It uses a try/catch to ensure the original default is restored even if the particular test fails.

krshrimali replied:
Awesome. Thanks @nairbv, I didn't know of that decorator.

nairbv commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,17 +53,29 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
 
 Tensor& acos_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, acos_stub); }
-Tensor acos(const Tensor& self) { return unary_op_impl(self, at::acos_out); }
+Tensor acos(const Tensor& self) { return unary_float_op_impl(self, at::acos_out); }
```
we should be able to handle the `_out` variants too.

mcarilli replied:
We can also split out `unary_op_impl_out` with a separate float-specific case.  This may make sense because that's the point where the `TensorIterator::unary_op` is constructed, so that's the point where we need to opt into or not opt into the PROMOTE behavior.

For what it's worth, I think `out=...` API calls are more closely related to the proposed `dtype=...` API calls than to the implicit promotion.  For the `out=...` and `dtype=...` variants, the user has requested that the result be a specific type.  Also, the `out=...`ness of an op variant is orthogonal to whether or not that op falls onto the implicit promote list.

Maybe it makes sense to have a single `unary_op_impl_out` with an optional `allow_promote=False` argument to serve all cases.  It isn't necessary for the present PR though. Do you think that should be part of this PR or a followup PR which adds explicit dtype flags?

nairbv replied:
I think it makes sense to include in this PR, it's more closely related to in-place, and I don't think actually affects the mixed-precision behavior you'll want to control with `dtype=`. 

The reason this seems so relevant to this PR is in `unary_float_op_impl` the `self.to(result.scalar_type())` will be incompatible with the out variants. e.g. if a user can:

```
a=torch.tensor(1, dtype=torch.half, device='cuda')
torch.log(a, out=a)
b=torch.tensor(1, dtype=torch.int)
torch.log(b)
```
Then they should also be able to:
```
torch.log(b, out=a)
```
If we disallow the last case I think we're exposing an inconsistent API. The last case (inside TensorIterator) will validate that the result of torch.log(int_tensor) can be cast to half (by checking torch.can_cast). By casting `self` to `result.scalar_type()` explicitly here we're bypassing the code that's supposed to handle this.

nairbv replied:
note that once implemented, an op like `torch.log(int_tensor, out=half_tensor)` will just be equivalent to (inside TensorIterator):
```
result=torch.log(int_tensor)
half_tensor.copy_(result)
```
Modifying the behavior to actually perform the operation at another precision (whether specified through `dtype=` or not) would still be a subsequent change.

mcarilli replied:
After discussion with @mruberry we're going to try option 1 first (allowing the `out=...` calls to opt-into PROMOTE).  @mruberry is that right or am i totally misinterpreting...

krshrimali replied:
Update: So we went ahead with the Option 1, after discussion with @nairbv, @mcarilli and @mruberry.

We allow `PROMOTE` for `out=...` calls as well as for unary floating ufuncs to avoid `self.to()` calls. For other cases, `result.scalar_type() != self.scalar_type()` condition in `unary_op_impl_out` definition will always return false, hence not allowing the iterator to opt-in to `PROMOTE` strategy.

krshrimali committed:
Avoid .to() call and use PROMOTE strategy for out variants & floating ufuncs

As per discussion on slack, this PR:

1. Avoids .to() call for floating ufuncs.
2. Opt-in to PROMOTE strategy for _out variants and unary floating ufuncs

Adding result.scalar_type() != self.scalar_type() allows only _out variants and unary floating ufuncs to opt-in to PROMOTE CommonDTypeStrategy, as those are the only 2 (currently) cases where self.dtype != result.dtype.

krshrimali committed:
Add support for all the ops; refine list; revisions in test code.

1. Add promotion support for all _out variants of all unary ufuncs, and for all unary "floating" ufuncs.
2. Revise list in test_torch.py
3. Use `@float_double_default_dtype` as suggested by @nairbv

As discussed with @mruberry, the test code needs work and should be converted into a test class. (TODO)

krshrimali commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -366,18 +366,18 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
-IMPLEMENT_UNARY_OP_VEC_CUDA(erfinv)
+IMPLEMENT_UNARY_FLOATING_OP_VEC(erf)
+IMPLEMENT_UNARY_FLOATING_OP_VEC(erfc)
+IMPLEMENT_UNARY_FLOATING_OP_VEC_CUDA(erfinv)
```
All the ops (`atan, cos, cosh, erf, erfc, exp, tan, tanh`) using `IMPLEMENT_UNARY_FLOATING_OP_VEC` will throw error for: `torch.atan(torch.tensor(2, dtype=torch.int8, device="cuda"))` , `RuntimeError: _th_atan_out not supported on CUDAType for Char`.

krshrimali replied:
Also, since all the ops now use new macros (`IMPLEMENT_UNARY_FLOATING_OP_VEC`, `IMPLEMENT_UNARY_FLOATING_OP_VEC_CUDA`), should we remove these (https://github.com/pytorch/pytorch/pull/33322/commits/0d1e167bd0c95c85f0d1fa1966ac3334edafdf14#diff-a249164f42f7e57a9a0028bbc604ae77R351-R358) macros?

krshrimali commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -62,7 +62,7 @@ static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl)
                   at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : \
                   at::empty({0}, self.options());
-  return out_impl(result, self.to(result.scalar_type()));
+  return out_impl(result, self);
```
The current implementation (using `PROMOTE` strategy for promotion in TensorIterator API for unary floating ufuncs) avoids `self.to(result.scalar_type())` call.

nairbv replied:
This approach seems a bit confusing to me. I can see why it works, but it works for an odd reason.

By relying on regular promotion logic inside tensor iterator, we're depending on the fact that `result_type(input_tensor, desired_result) == desired_result` for any possible dtype. This seems to work -- `torch.result_type(int_tensor, float32_tensor)` gives us float32 not float64... but this is relying on the table in ScalarType.h to get the result, not actually relying directly on the call to `torch.get_default_dtype`. 

I think I'd prefer to be more explicit in how we handle this, to simplify any changes that might be needed for additional dtypes later.

krshrimali commented on test/test_torch.py:
```diff
@@ -34,7 +34,7 @@
 import torch.backends.quantized
 import torch.testing._internal.data
-
+from test_type_promotion import float_double_default_dtype
```
To use `float_double_default_dtype`, this import had to be added.

krshrimali committed:
remove dead/unused macros

Commit:
example unary floating ufunc impl

Commit:
test_torch.py improvements

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
Merge remote-tracking branch 'upstream/unary_floating_ufunc' into int-to-float-unary-ops

krshrimali committed:
Add ops support, delete previous changes & merge conflicts

krshrimali committed:
Register floating ufuncs in test class for floating ufuncs

krshrimali committed:
minor typo

krshrimali committed:
Solve minor typos

krshrimali committed:
ComplexFloat not supported for some ops. Use complex_on=()

krshrimali commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,112 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+    'erf': ufunc_meta(is_floating=True, complex_on=()),
+    'erfc': ufunc_meta(is_floating=True, complex_on=()),
+    'erfinv': ufunc_meta(is_floating=True, complex_on=()),
+    'lgamma': ufunc_meta(is_floating=True, complex_on=())
```
Ops which use `complex_on=()` are those which do not support Complex Type (`torch.complex64`, `torch.complex128`) Tensors on CPU and CUDA devices. By default, `complex_on` is set to `('cpu', 'cuda')`.

krshrimali committed:
Adds ops list for onnx tests

krshrimali committed:
Add tests for out= and inplace calls

krshrimali commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,138 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            self.assertEqual(op(t).dtype, in_type)
+
+        only_cpu_ops = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
```
These ops do not support type promotion for `out=` calls for `cuda` devices. That's why the check later (16413 line).

mruberry replied:
What happens when they're called with out= on CUDA?

krshrimali replied:
For `out=` calls on CUDA (with input tensors as Int Tensors), it throws this error: `RuntimeError: _th_atan_out not supported on CUDAType for Char`. In my opinion, this is because the `only_cpu_ops` (to be renamed to: `only_cpu_out_ops`, like atan, cos, cosh etc.) do not implement through `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_VEC_CUDA` which has `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_OUT_INPLACE(op, cuda, CUDA)` in it's call.

Indirectly, these cpu-out-only ops only implement the `out=` call for CPU devices, I assume!

mruberry replied:
Interesting. Thanks for the clarification.

krshrimali commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,138 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+            op_inplace = getattr(torch, op_str + "_")
+        except AttributeError:
+            # lgamma, digamma and erfinv don't have in-place functions
+            return
```
This check ensures no AttributeError is raised for `digamma, erfinv, lgamma` ops. These do not have in-place overloads.

mruberry replied:
That look really good.

krshrimali replied:
Thanks, @mruberry! :slightly_smiling_face:

krshrimali committed:
Add support for mvlgamma, minor linter changes

mvlgamma calls lgamma_ (in-place call, which does not perform floating ufunc promotion).
mvlgamma checks input tensor type (anything except FloatingType is thrown as error).

In order to promote IntegralType input tensors, and pass through the check, this commit makes the required changes!

krshrimali committed:
correct onnx tests for erf, floor, ceil

krshrimali commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -341,17 +341,19 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
 Tensor mvlgamma(const Tensor& self, int64_t p) {
-  mvlgamma_check(self, p);
-  Tensor args = native::arange(-p / 2. + 0.5, 0.5, 0.5, self.options());
+  mvlgamma_check(self, p, /*floating_ufunc=*/true);
```
`true` is passed to ensure `mvlgamma_check` does not go through the `TORCH_CHECK(at::isFloatingType(self.scalar_type()), "mvlgamma is not implemented for ", self.scalar_type())` check.

mruberry replied:
See above note: should we change mvlgamma at all?

krshrimali commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -341,17 +341,19 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+static inline void mvlgamma_check(const Tensor& self, int64_t p, bool floating_ufunc=false) {
+  if(!floating_ufunc)
+    TORCH_CHECK(at::isFloatingType(self.scalar_type()),
+            "mvlgamma is not implemented for ", self.scalar_type());
```
This check could have been removed, but `mvlgamma_` (inplace call) should not support type promotion.

mruberry replied:
What if we don't make any changes to mvlgamma?

krshrimali replied:
I think, it will be nice to have changes to mvlgamma:

`torch.mvlgamma(torch.tensor((1,), dtype=torch.int8), p=2)` throws `RuntimeError: mvlgamma is not implemented for Char` error. 
`torch.mvlgamma(torch.tensor((1,), dtype=torch.float32), p=2)` returns `tensor([1.447])`. 

Based on the above comparison, I believe `mvlgamma` should have this promotion? I unfortunately don't have contextual knowledge about this!

mruberry replied:
Let's leave it alone for now. We can always change it later.

krshrimali commented:
ONNX tests fail because of existing `test_log` in `TestONNXRuntime` (error: `AssertionError: test_log already in TestONNXRuntime`). `test_log1p`, `test_rsqrt` and `test_log` are already present in TestONNXRuntime. 

@mruberry - Should these ops be removed from the list (https://github.com/pytorch/pytorch/blob/76d3b3510c39b6979d6a211f1239cf422a6c0a62/test/onnx/test_pytorch_onnx_onnxruntime.py#L3114) or should the tests (e.g. https://github.com/pytorch/pytorch/blob/76d3b3510c39b6979d6a211f1239cf422a6c0a62/test/onnx/test_pytorch_onnx_onnxruntime.py#L2559) be removed? 

In my opinion, the tests (for log1p, rsqrt and log) can be removed as it is exactly what our function (https://github.com/pytorch/pytorch/blob/76d3b3510c39b6979d6a211f1239cf422a6c0a62/test/onnx/test_pytorch_onnx_onnxruntime.py#L3146) does!

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,18 +53,81 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
+// produces non-integral values. Implementers should continue to use the other unary functions, like unary_op_impl_,
+// since they explicitly specify the output type.
+template <typename OutImpl>
+static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl) {
```
Is this still in use?

krshrimali replied:
No, sorry. I'll remove this in the next commit!

krshrimali replied:
This should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/0a3ca09ff29927b82399d858918dca69a799c085. Thanks! :)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -53,18 +53,81 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
+  TORCH_CHECK(!isIntegralType(result.scalar_type(), /*includeBool=*/ true),
+    "Attempted to call a function that returns a tensor of floating dtype ",
+    "but requested a tensor of bool or integral dtype. ",
+    "Specify an out tensor with a floating dtype or a floating dtype argument.");
```
Let's kill this last sentence. It doesn't add anything.

krshrimali replied:
Sure. Done! :) Should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/19d680b7d20f993009f475e272e5e5c1cc75489f.

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,6 +3109,58 @@ def setup_rnn_tests():
+# Unary floating ufunc tests
+# Tests that integral arguments are properly cast to float
+unary_floating_ufuncs = (
+    'cos',
```
Reorder this list logically. For example, you can start with the trig functions, and put a few on the same line: `'cos', 'sin', 'tan'`...

krshrimali replied:
Agreed! I reordered the list (https://github.com/pytorch/pytorch/pull/33322/commits/174d85e75be51508760b0406ef21cccdd462e18d).

Here is the revised list:

```
unary_floating_ufuncs = (
    'sin', 'cos', 'tan',
    'asin', 'acos', 'atan',
    'sinh', 'cosh', 'tanh',
    'ceil', 'floor',
    'exp', 'expm1', 'log', 'log10', 'log1p', 'log2',
    'sqrt', 'rsqrt',
    'trunc',
    'erf', 'erfc', 'erfinv',
    'lgamma', 'digamma',
    'sigmoid'
)
```

Let me know if this requires further changes!

krshrimali replied:
I took the liberty of reordering list in `test_torch.py` as well (https://github.com/pytorch/pytorch/pull/33322/commits/0b0fdef7ec56ae1ebdc13bb93656d575065e9218). The order is similar as above for consistency.

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,6 +3109,58 @@ def setup_rnn_tests():
+    'digamma',
+    'rsqrt',
+    'sigmoid',
+    'acos'
```
Is this list complete?

krshrimali replied:
Yes. I cross-checked the list. This list looks complete to me (except `mvlgamma` - which is under discussion).

mruberry commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+
+        only_cpu_ops = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
+        for in_type in my_float_types + int_types:
+            if(op_str in only_cpu_ops and self.device_type == 'cuda'):
```
if <space> (

krshrimali replied:
Thanks! Should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/19d680b7d20f993009f475e272e5e5c1cc75489f.

mruberry commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            self.assertEqual(op(t).dtype, in_type)
+
+        only_cpu_ops = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
```
Give this a more descriptive name, like "only_out_on_cpu"

krshrimali replied:
Thanks, this sounds reasonable! (https://github.com/pytorch/pytorch/pull/33322/commits/0c9463aabee3a0449cdeae9560b14ab5a7ef5f20).

mruberry commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+                continue
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            for out_type in my_float_types:
+                if(in_type == torch.bool and out_type in complex_types):
```
What happens when PyTorch tries to promote bool to complex?

krshrimali replied:
I think this is because of the common dtype calculation in the TensorIterator logic. `torch.cos(BoolTensor, out=<tensor with complex64>)` throws `RuntimeError: dtype != ScalarType::Undefined INTERNAL ASSERT FAILED at /home/krshrimali/Documents/pytorch/pytorch/aten/src/ATen/native/TensorIterator.cpp:124, please report a bug to PyTorch. `.

krshrimali replied:
I thought to dig further on this error.

For the common dtype calculation, the function `promote_skip_undefined` (at: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TypeProperties.cpp#L64) uses promotion table in PyTorch (https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h#L438). 

Here, for bool (`b1`) to complex types (`c2`, `c4`, `c8`), the output type is `ud` (undefined). And then we have an assert (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L124) which throws this error.

mruberry commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+                self.assertRaises(RuntimeError, lambda: op(t, out=out_t))
+
+        try:
+            op_inplace = getattr(torch, op_str + "_")
```
Interesting that you can getattr from torch here. I would have thought you'd have to use getarr(torch.Tensor...) ?

krshrimali replied:
To get in-place calls, I had to separately use `getattr(torch, <op_inplace_name>)` here. It works fine normally in Python as well:

```
getattr(torch, "atan_")(torch.tensor((1,), dtype=torch.float32))
```

Are you expecting anything else instead of this method?

mruberry replied:
`op_inplace = gettattr(torch.Tensor, op_str + "_")`

krshrimali committed:
refine TORCH_CHECK statement, remove dead code

krshrimali committed:
nit: if ( instead of if(

krshrimali committed:
rename only_cpu_ops to only_out_on_cpu

krshrimali committed:
reorder the uanry floating ufuncs list logically

krshrimali committed:
reorder lists for unary floating ufuncs

krshrimali committed:
Add comments where required in tests

krshrimali committed:
Add comment for helper functions used

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,6 +3109,42 @@ def setup_rnn_tests():
+        self.run_test(module(), (x,))
+        self.run_test(torch.jit.trace(module(), (x,)), (x,))
+
+    test_name = "test_" + op_name
```
Let's change the test names to avoid conflicts. Maybe "test_<op_name>_ufunc"?

krshrimali replied:
Done, thanks! https://github.com/pytorch/pytorch/pull/33322/commits/49ec5a40809f4f496b2850ab6a45579f92c8f520

mruberry commented on test/test_torch.py:
```diff
@@ -16330,16 +16329,148 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+                # and output as complex tensors
+                if (in_type == torch.bool and out_type in complex_types):
+                    continue
+                out_t = torch.tensor((), device=device, dtype=out_type)
```
Should this have a size of 1? torch.empty((1,), ...)?

krshrimali replied:
Done! https://github.com/pytorch/pytorch/pull/33322/commits/49ec5a40809f4f496b2850ab6a45579f92c8f520

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,6 +3109,42 @@ def setup_rnn_tests():
+            def forward(self, x):
+                return op(x)
+
+        x = (torch.rand(2, 3, 4) * 10).to(torch.int)
```
Let's switch torch.int to torch.long. That was my mistake.

krshrimali replied:
Sure! Should be resolved here: https://github.com/pytorch/pytorch/pull/33322/commits/49ec5a40809f4f496b2850ab6a45579f92c8f520.

krshrimali committed:
Revert changes for mvlgamma; change name of generated tests in ONNX

krshrimali committed:
Use torch.empty and torch.long (in place of torch.int)

Commit:
Register floating ops in symbolic_opset9

krshrimali committed:
Remove ops which aren't supported by ONNX

krshrimali committed:
Unary Floating Ufuncs do not support complex types with cuda devices

krshrimali committed:
Merge branch 'int-to-float-unary-ops' of http://github.com/krshrimali/pytorch into int-to-float-unary-ops

krshrimali committed:
Tests for float to complex conversion; correct complex_on argument

krshrimali committed:
Add argument complex_out_cpu to metadata for better readability

krshrimali committed:
erf does not support opset version below 9

Commit:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
min opset version 9 for Sinh and Cosh

krshrimali commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3136,7 +3136,7 @@ def forward(self, x):
     test_name = "test_" + op_name + "_ufunc"
     assert not hasattr(cls, test_name), "{0} already in {1}".format(test_name, cls.__name__)
-    if(op_name in ['erf']):
+    if(op_name in ['erf', 'sinh', 'cosh']):
```
This is because of the following errors (`caffe2_onnx_ort1_py3_6_clang7_ubuntu16_04_test`):

```
RuntimeError: No Op registered for Cinh with domain_version of 7
RuntimeError: No Op registered for Cinh with domain_version of 8
RuntimeError: No Op registered for Sinh with domain_version of 7
RuntimeError: No Op registered for Sinh with domain_version of 8
```

My take:

From https://github.com/onnx/onnx/blob/master/docs/Operators.md#Sinh and https://github.com/onnx/onnx/blob/master/docs/Operators.md#Sinh, it states:

> This version of the operator has been available since version 9 of the default ONNX operator set.

I assume the reason it gives segmentation fault for Opset7 and Opset8 is that it does not support opset versions below 9 (https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py#L737). This is based on my limited understanding of skipIfUnsupportedMinOpsetVersion from https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_common.py#L45.

Commit:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

Commit:
Revise ops list after #34093 is merged (support for complex type promotion)

Commit:
Merge branch 'int-to-float-unary-ops' of http://github.com/krshrimali/pytorch into int-to-float-unary-ops

krshrimali committed:
Merge branch 'int-to-float-unary-ops' of https://github.com/krshrimali/pytorch into int-to-float-unary-ops

krshrimali commented on test/test_torch.py:
```diff
@@ -16454,38 +16454,38 @@ def test_svd_tall_all_col_maj(self, device, dtype):
+    'erfinv': ufunc_meta(is_floating=True, complex_on=(), complex_out=()),
+    'lgamma': ufunc_meta(is_floating=True, complex_on=(), complex_out=()),
+    'digamma': ufunc_meta(is_floating=True, complex_on=(), complex_out=()),
+    'sigmoid': ufunc_meta(is_floating=True, complex_on=('cpu'), complex_out=('cpu'))
```
These changes in the `complex_out` arguments are after the PR #34093 (which adds type promotion logic for complex numbers) is merged. 

Unary Floating Ufuncs which do not support complex tensors as input/out argument on:

* CPU devices: `expm1, log1p, erf, erfc, erfinv, lgamma, digamma`.
* CUDA devices: `cos, tan, tanh, cosh, atan, exp, expm1, log1p, erf, erfc, erfinv, lgamma, digamma, sigmoid`.

krshrimali commented on test/test_torch.py:
```diff
@@ -16534,26 +16534,27 @@ def test_fn(self, device):
-                    continue
-                out_t = torch.empty((1,), device=device, dtype=out_type)
-                self.assertEqual(op(t, out=out_t).dtype, out_type)
+                    self.assertRaises(RuntimeError, lambda: op(t, out=out_t))
```
Adding `self.assertRaises` is a better strategy for test cases instead of using `continue` and ignoring the calls, IMO.

mruberry replied:
Agreed.

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

mruberry commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1222,7 +1230,7 @@ def abs(g, self):
+    return sym_help._unary_ufunc_helper(g, self, "Log")
 
 
 def log1p(g, self):
```
log1p is a unary floating ufunc on your list but the ONNX export isn't modified. Is it working as expected?

krshrimali replied:
I didn't modify, assuming it calls `log` already. The same reason we don't need to modify `rsqrt`? https://github.com/pytorch/pytorch/blob/e0373d504746f2d5d0b0187e4b51f3a67b7f9e63/torch/onnx/symbolic_opset9.py#L217

All the tests pass though (and `log1p` is listed in the list in test_pytorch_onnx_onnxruntime.py file).

Should log1p export be modified? I'm not sure why and how.

mruberry replied:
SGTM.

mruberry commented on torch/onnx/symbolic_helper.py:
```diff
@@ -375,6 +375,21 @@ def _is_all_integral(scalars):
+        dtype = torch.get_default_dtype()
+        assert dtype in torch_float_dtype_to_onnx.keys()
+        onnx_scalar_type = torch_float_dtype_to_onnx[dtype]
+        self = g.op("Cast", self, to_i=onnx_scalar_type)
```
I think I found a simpler way to generate this cast that doesn't require the table above. Try:

`self = g.op("Cast", input, to_i=scalar_type_to_pytorch_type.index(torch.get_default_dtype()))`

krshrimali replied:
Sure! Let me try if it works fine.

krshrimali replied:
I think you meant:

```
type = scalar_type_to_pytorch_type.index(torch.get_default_dtype())
self = g.op("Cast", self, to_i=scalar_type_to_onnx[type])
```

Please correct me if I'm wrong.

krshrimali replied:
This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,12 +3109,48 @@ def setup_rnn_tests():
+    'asin', 'acos', 'atan',
+    'sinh', 'cosh', 'tanh',
+    'ceil', 'floor',
+    'exp', 'log', 'log1p', 'log2',
```
log1p and log2 are (correctly) in this list, but their ONNX translations haven't been changed. Were they already working correctly?

krshrimali replied:
I believe so. Since they call other unary floating ufunc which have been changed, so IMO: they don't need the modifications. Please let me know if you think they should be changed. The tests did pass successfully for all the ops.

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3109,12 +3109,48 @@ def setup_rnn_tests():
+
+    test_name = "test_" + op_name + "_ufunc"
+    assert not hasattr(cls, test_name), "{0} already in {1}".format(test_name, cls.__name__)
+    if(op_name in ['erf', 'sinh', 'cosh']):
```
`if op_name in [...]:`

would be more pythonic

krshrimali replied:
Agreed! This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+
+# Dict of unary ufuncs and their associated test metadata
+unary_ufuncs = {
+    'sin': ufunc_meta(is_floating=True, complex_on=('cpu')),
```
Is complex_on ever distinct from complex_out? If not, remove complex_out?

krshrimali replied:
For ops like `log, log10, log2, sqrt, rsqrt, trunc` it's distinct. I think we'll have to keep it?

krshrimali replied:
I was wrong. We don't need `complex_out`. Removed in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+_ufunc_meta = collections.namedtuple('_ufunc_meta',
+                                     'is_floating alt_impl complex_on complex_out')
+def ufunc_meta(is_floating=False, alt_impl=None, complex_on=('cpu', 'cuda'), complex_out=('cpu', 'cuda')):
+    return _ufunc_meta(is_floating, alt_impl, complex_on, complex_out)
```
Let's remove alt_impl for now, since it won't be used by this PR. (The future PR I included it for can add it back.)

krshrimali replied:
Sure! Sounds good.

krshrimali replied:
This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+# the default scalar type.
+def generate_unary_floating_ufunc_promo_test(cls, op_str):
+    int_types = [torch.bool, torch.uint8, torch.int8, torch.int16,
+                 torch.int, torch.long]
```
torch.int32, torch.int64 for consistency (this was my mistake, sorry!)

krshrimali replied:
This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+            self.assertEqual(op(t).dtype, in_type)
+
+        # These ops only support out= calls on CPU devices
+        only_out_on_cpu = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
```
I don't think this list is correct. The following program runs for me:

```
import torch

device = 'cuda'
t = torch.tensor((1,), device=device, dtype=torch.float)
e = torch.empty((1,), device=device, dtype=torch.float)
torch.cos(t, out=e)
print(e)
```

Also, function metadata like this should be put into ufunc_meta and queried, like "complex_on"

krshrimali replied:
You are right, this list is not needed. Will be removed in the next commit.

krshrimali replied:
This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+        for in_type in my_float_types + int_types:
+            # Floating type promotion for out= calls on ops listed in only_out_on_cpu
+            # not supported on cuda devices
+            if (op_str in only_out_on_cpu and self.device_type == 'cuda'):
```
This query will need to change when the above list is modified

krshrimali replied:
Done! This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+                out_t = torch.empty((1,), device=device, dtype=out_type)
+                self.assertRaises(RuntimeError, lambda: op(t, out=out_t))
+
+        # Test for out= code paths (with input as float tensors and out as complex tensors)
```
Isn't this test redundant with the previous one?

krshrimali replied:
Yes, you are right. This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+    default_types_cuda = [torch.half, torch.float, torch.double]
+
+    def test_fn(self, device):
+        op = getattr(torch, op_str)
```
We should test the methods, too: 

`meth = getattr(torch.Tensor, op_str)`

krshrimali replied:
Agreed. This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+            for in_type, def_type in product(int_types, my_default_types):
+                torch.set_default_dtype(def_type)
+                t = torch.tensor((1,), device=device, dtype=in_type)
+                self.assertEqual(op(t).dtype, def_type)
```
For method just add:

`self.assertEqual(meth(t).dtype, def_type)`

krshrimali replied:
Done! :) This is solved in the recent commit: https://github.com/pytorch/pytorch/pull/33322/commits/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -16448,16 +16447,160 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+
+        for in_type in my_float_types:
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            self.assertEqual(op(t).dtype, in_type)
```
Add analogous method test here, too.

krshrimali replied:
Done! https://github.com/pytorch/pytorch/commit/29ae207f362efe4040b598355cfe1db1fadb82ad. Thanks! :)

nairbv commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -58,13 +58,66 @@ static inline Tensor& unary_op_impl_(Tensor& self, OutImpl& out_impl) {
+static inline Tensor unary_floating_ufunc_op_impl(const Tensor& self, OutImpl& out_impl) {
+  if (isIntegralType(self.scalar_type(), /*includeBool=*/ true)) {
+    const auto scalar_type = typeMetaToScalarType(c10::get_default_dtype());
+    Tensor result = at::empty({0}, self.options().dtype(scalar_type));
```
I think you can just use `{}` here to avoid allocating space for the 0.

gchanan replied:
not sure what you mean by allocating space for the 0, but `{}` would allocate a 1-element tensor.

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

Commit:
Add support for angle; Revise tests, code reviews

Commit:
Merge branch 'int-to-float-unary-ops' of http://github.com/krshrimali/pytorch into int-to-float-unary-ops

krshrimali committed:
Resolve ONNX Test Failures

krshrimali committed:
Update JIT shape analysis for unary floating funcs

This commit also removes the unary floating ops from simple_unary_ops, 
adds those in new list unary_floating_ufuncs.

krshrimali commented on torch/onnx/symbolic_helper.py:
```diff
@@ -376,6 +376,14 @@ def _is_all_integral(scalars):
+def _unary_ufunc_helper(g, self, op_name):
+    if not _is_fp(self):
+        type = scalar_type_to_pytorch_type.index(torch.get_default_dtype())
+        self = g.op("Cast", self, to_i=scalar_type_to_onnx[type])
```
As per review suggestion from @mruberry : https://github.com/pytorch/pytorch/pull/33322#discussion_r395487581.

krshrimali commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -776,6 +776,58 @@ class ShapePropagator {
+        }
+      return type_vec_t{};
+      }};
+
```
(From on-slack conversation with @mruberry) Since type promotion is supported for unary floating ufuncs, this had to be updated. A new list `unary_floating_ufuncs` has been added with the unary floating ops, and updated lambda function.

krshrimali commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -787,15 +839,9 @@ class ShapePropagator {
     static const register_formula_for simple_unary_ops{
         {
             "aten::abs(Tensor self) -> Tensor",
-            "aten::acos(Tensor self) -> Tensor",
```
Unary Floating Ufuncs have been moved from this list to the new list (unary_floating_ufuncs).

krshrimali committed:
Minor typo, tests should pass now

krshrimali committed:
typo solved, aten::erf

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

mcarilli commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -58,13 +58,66 @@ static inline Tensor& unary_op_impl_(Tensor& self, OutImpl& out_impl) {
+  if (isIntegralType(self.scalar_type(), /*includeBool=*/ true)) {
+    const auto scalar_type = typeMetaToScalarType(c10::get_default_dtype());
+    Tensor result = at::empty({0}, self.options().dtype(scalar_type));
+    return out_impl(result, self.to(scalar_type));
```
out of the loop for a bit fighting with CI for autocasting... @mruberry git blame (or credit!) says it's your work (from a [merged branch](https://github.com/pytorch/pytorch/commit/fa0ea9a49a1353ac996f88ab624591c520a91a61)) to split `unary...impl` into two versions.  The other version (`unary_floating_ufunc_op_impl`), for integer tensor input, creates result as the default type and sends self and result along to TensorIterator as they are.  This version precasts self to match the output type, so TensorIterator doesn't need to handle mismatched types internally.  

Splitting those two versions is a good way to deal with the fact that some backend functions may handle mismatched self+result types and some may not.  Afaict `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_VEC` macros are the only place `unary_floating_ufunc_cast_op_impl` is used, ie, the ops handled by those macros are the ones you decided should receive precasting.

I'd like to make sure I understand how you and @krshrimali decided which ops receive precasting.  The ops handled by `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_VEC` match the ops from [this](https://github.com/pytorch/pytorch/pull/33322#discussion_r394318659) (possibly outdated) comment, with the addition of `erfinv` and `lgamma`.  If i'm reading that comment correctly, the ops listed there don't support internal int->default_type promotion when called on CUDA tensors, which is a reasonable criterion to route them through `unary_floating_ufunc_cast_op_impl` rather than `unary_floating_ufunc_op_impl`.

However (if im reading it correctly) those same ops _do_ support internal promote when called with CPU inputs.  Right now the `_impl` each op calls is based only on the op itself.  Since internal promote support depends on both the op and `self`'s device, do you think it makes sense to have `unary_floating_ufunc_cast_op_impl` choose not to precast if `is_cuda()` is false?

side note:  `unary_floating_ufunc_cast_op_impl` only precasts integer inputs, in other words, floating point inputs never precast no matter which `_impl` they route through, and always rely on internal TensorIterator promotion.  That's great for GPU performance, and will be great for performance with the hoped-for output `dtype=...` argument as well.

krshrimali replied:
I agree to what you suggest here @mcarilli. For the functions using the helper macros, adding a simple if-else condition works fine with all the tests:

```cpp
#define IMPLEMENT_UNARY_FLOATING_UFUNC_OP_CORE(op)                     \
  Tensor op(const Tensor& self) {                                      \
    if (!self.is_cuda) \
      return unary_floating_ufunc_op_impl(self, at::op##_out); \
    else \
      return unary_floating_ufunc_cast_op_impl(self, at::op##_out);      \
  }
```

@mruberry - What are your views?

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

mruberry commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -776,6 +776,58 @@ class ShapePropagator {
+      },
+      [](Node* node) -> type_vec_t {
+        if (auto type = node->input(0)->type()->cast<TensorType>()) {
+          auto ret = type->withDim(0);
```
Do you need this withDim? It doesn't appear in the simple_unary_ops lambda.

krshrimali replied:
You are right. Thank you!

krshrimali committed:
Remove withDim(0) call

krshrimali committed:
minor changes

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into int-to-float-unary-ops

krshrimali committed:
Add angle signature to tests

krshrimali committed:
Add angle support in UnaryOps.cpp

krshrimali committed:
Avoid AttributeErrors for in-place methods (angle exception)

krshrimali committed:
angle_cuda not implemented for torch.half

krshrimali commented on test/test_torch.py:
```diff
@@ -16714,6 +16714,10 @@ def test_fn(self, device):
+            # angle_cuda is not implemented for torch.half
+            if op_str == 'angle':
+                my_default_types = [type for type in default_types_cuda if type != torch.half]
+                my_float_types = [type for type in float_types_cuda if type != torch.half]
```
`angle_cuda` is not implemented for `torch.half`. For reference, see:

1. `AT_DISPATCH_ALL_TYPES_AND_COMPLEX` used <a href="https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UnaryComplexKernels.cu#L26">here</a> in the definition of `angle_kernel_cuda`.
2. Definition of the macro <a href="https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L297">here</a>.

krshrimali commented on test/test_torch.py:
```diff
@@ -16651,16 +16650,154 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+        try:
+            op_inplace = getattr(torch.Tensor, op_str + "_")
+        except AttributeError:
+            return
```
This is to avoid any `AttributeError` like in case of `angle_` (in-place op for angle is not implemented).

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -86,80 +139,80 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) { return unary_op_im
+Tensor expm1(const Tensor& self) { return unary_floating_ufunc_op_impl(self, at::expm1_out); }
+Tensor& expm1_(Tensor& self) { return unary_floating_ufunc_op_impl_(self, at::expm1_out); }
 
 Tensor& frac_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, frac_stub); }
```
frac is a composite that can alternatively be implemented as 

`(self - torch.floor(torch.abs(self)) * torch.sign(self))`

The call to floor should promote to floating. Seems like frac should be a unary floating ufunc, too.

krshrimali replied:
Added frac as unary floating ufunc. It's not there in the ONNX operator lists (https://github.com/onnx/onnx/blob/master/docs/Operators.md) - so no changes needed there. Also added in the JIT `shape_analysis.cpp` file as unary floating ufunc.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -303,45 +356,41 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
-    Tensor result = at::empty({0}, self.options());                    \
-    at::op##_out(result, self);                                        \
-    return result;                                                     \
+    return unary_floating_ufunc_cast_op_impl(self, at::op##_out);      \
```
This still needs to be updated to cast if not CPU. The query should be if (CPU) ... else ...

krshrimali replied:
Done! https://github.com/pytorch/pytorch/pull/33322/commits/71e23c673e902017ab6d86445ace11bb1007144e

mruberry commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -776,6 +776,57 @@ class ShapePropagator {
+      {
+        "aten::acos(Tensor self) -> Tensor",
+        "aten::angle(Tensor self) -> Tensor",
+        "aten::tanh(Tensor self) -> Tensor",
```
tanh appears twice

krshrimali replied:
Sorry, my bad. Fixed here: https://github.com/pytorch/pytorch/pull/33322/commits/71e23c673e902017ab6d86445ace11bb1007144e

krshrimali committed:
Add frac, cast logic to not happen on CPU

krshrimali committed:
Register frac in unary floating ufunc in jit shape_analysis

mruberry commented on test/test_torch.py:
```diff
@@ -16651,16 +16650,155 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+
+        # Test for out= code paths
+        for in_type in my_float_types + int_types:
+            if self.device_type not in unary_ufuncs[op_str].complex_on:
```
Isn't this break premature? That is, isn't this saying if you don't support complex we should test the out= code path at all?

krshrimali commented:
Closing this since this was taken up in another PR. Thank you @mcarilli @mruberry for being patient and helping me with this PR. Really appreciate it, apologies that it couldn't make it to the upstream.
